### PR TITLE
feat: add create-admin action

### PR DESCRIPTION
**Pull request by amandahla**
Applicable spec: <link>

### Overview

This PR adds a action create-admin to the Maubot charm.

Adding a user to Maubot means adding a new item to the "admins" in the configuration file. 
Then the charm needs to restart Maubot for the changes to take effect.

- Action outputs a password.
- Action fails if name is "root" (reserved).

### Rationale

Create admin Maubot user.

### Juju Events Changes

Add new action create-admin.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

No CH available.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -43,49 +45,57 @@ def __init__(self, *args: typing.Any):
...
+        self.framework.observe(self.on.create_admin_action, self._on_create_admin_action)
         # Integrations events handlers
         self.framework.observe(self.postgresql.on.database_created, self._on_database_created)
         self.framework.observe(self.postgresql.on.endpoints_changed, self._on_endpoints_changed)
         self.framework.observe(self.ingress.on.ready, self._on_ingress_ready)
         self.framework.observe(self.ingress.on.revoked, self._on_ingress_revoked)
 
-    def _configure_maubot(self, container: ops.Container) -> None:
-        """Configure maubot.
+    def _get_configuration(self) -> dict:
```
Comment (arturo-seijas) [inline src/charm.py:63]: Perhaps you can add more specific return types?

File: src/charm.py [hunk 2]
```diff
@@ -104,6 +114,31 @@ def _on_ingress_revoked(self, _: IngressPerAppRevokedEvent) -> None:
...
+        name = event.params["name"]
+        if name == "root":
+            event.fail("root is reserved, please choose a different name")
+            return
+        if (
+            not self.container.can_connect()
+            or MAUBOT_NAME not in self.container.get_plan().services
+            or not self.container.get_service(MAUBOT_NAME).is_running()
+        ):
+            event.fail("maubot is not ready")
```
Comment (arturo-seijas) [inline src/charm.py:133]: The message does not seem very user friendly for
`MAUBOT_NAME not in self.container.get_plan().services` and even `not self.container.get_service(MAUBOT_NAME).is_running()`

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -63,3 +63,28 @@ async def test_build_and_deploy(
     )
     assert response.status_code == 200
     assert "Maubot Manager" in response.text
+
+
+async def test_create_admin_action(ops_test: OpsTest):
```
Comment (arturo-seijas) [inline tests/integration/test_charm.py:68]: How about some tests for the unhappy paths?

File: src/charm.py [hunk 4]
```diff
@@ -104,6 +114,31 @@ def _on_ingress_revoked(self, _: IngressPerAppRevokedEvent) -> None:
...
 
+    # Actions events handlers
+    def _on_create_admin_action(self, event: ops.ActionEvent) -> None:
+        """Handle delete-profile action.
+
+        Args:
+            event: Action event.
+        """
+        name = event.params["name"]
+        if name == "root":
```
Comment (jdkandersson) [inline src/charm.py:125]: Should we also check for whether the user already exists?

File: src/charm.py [hunk 5]
```diff
@@ -104,6 +114,31 @@ def _on_ingress_revoked(self, _: IngressPerAppRevokedEvent) -> None:
         """Handle ingress revoked event."""
         self._reconcile()
 
+    # Actions events handlers
+    def _on_create_admin_action(self, event: ops.ActionEvent) -> None:
```
Comment (jdkandersson) [inline src/charm.py:118]: Do you think we also need to support deleting a user or changing the password for a user?

File: tests/integration/test_charm.py [hunk 6]
```diff
@@ -63,3 +63,28 @@ async def test_build_and_deploy(
...
+
+    assert "password" in action.results
+    password = action.results["password"]
+    response = requests.post(
+        "http://127.0.0.1/_matrix/maubot/v1/auth/login",
+        timeout=5,
+        headers={"Host": "maubot.local"},
+        data=f'{{"username":"{name}","password":"{password}"}}',
+    )
+    assert response.status_code == 200
```
Comment (jdkandersson) [inline tests/integration/test_charm.py:89]: Nice that you check that the user is working!

File: src/charm.py (hunk 2 above)
Comment (amandahla) [reply src/charm.py:133]: Could you give me more details?
Do you mean that "maubot is not ready" does not correspond to what is being checked?

File: src/charm.py (hunk 5 above)
Comment (amandahla) [reply src/charm.py:118]: Deleting a user might be handy, I'll work on that in a further PR.

File: src/charm.py (hunk 1 above)
Comment (amandahla) [reply src/charm.py]: Sure, changed.

File: tests/integration/test_charm.py (hunk 3 above)
Comment (amandahla) [reply tests/integration/test_charm.py]: Good suggestion, added.

File: src/charm.py (hunk 4 above)
Comment (amandahla) [reply src/charm.py:138]: Good catch! Added.

File: src/charm.py (hunk 2 above)
Comment (arturo-seijas) [reply src/charm.py]: Well, I don't remember anymore. It looks fine today

**Review by nrobinaubertin — DISMISSED** [review]

File: src/charm.py [hunk 7]
```diff
@@ -104,6 +114,48 @@ def _on_ingress_revoked(self, _: IngressPerAppRevokedEvent) -> None:
         """Handle ingress revoked event."""
         self._reconcile()
 
+    # Actions events handlers
+    def _fail_event(self, event: ops.ActionEvent, results: Dict[str, str], message: str) -> None:
```
Comment (javierdelapuente) [inline src/charm.py:118]: Is results necessary? It is always a dictionary with two keys and empty values in them. Maybe construct it in this function?

Comment (amandahla) [reply src/charm.py:118]: I'm planning to add more actions where this can be used as well.

Comment (javierdelapuente) [reply src/charm.py:122]: Maybe adding it in the next PR when it is required (and maybe with a different signature)?

Comment (amandahla) [reply src/charm.py]: I thought about it..WDYT about a custom exception? So there is no need for the function anymore.

**Review by javierdelapuente — APPROVED** [review]

**Review by jdkandersson — COMMENTED** [review]
Broadly looks good, I'll leave the approval to @gregory-schiano

**Review by nrobinaubertin — APPROVED** [review]

**Pull request merged**